kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt

```diff
@@ -203,4 +204,55 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
     val options = JsonReader.Options.of(*bindings.map { it?.name ?: "\u0000" }.toTypedArray())
     return KotlinJsonAdapter(constructor, bindings, options).nullSafe()
   }
+
+  private fun jsonAdapter(type: KType, moshi: Moshi, qualifierAnnotations: Set<Annotation>):
+      JsonAdapter<Any?> {
+    val javaType = type.javaType
+    val javaRawType = javaType.rawType
+    val create: () -> MutableCollection<Any?>
+    if (javaRawType == List::class.java || javaRawType == Collection::class.java) {
+      create = { ArrayList() }
+    } else if (javaRawType == Set::class.java) {
+      create = { LinkedHashSet() }
+    } else {
+      return moshi.adapter<Any>(javaType, qualifierAnnotations)
+    }
+    if (!qualifierAnnotations.isEmpty()) {
+      return moshi.adapter<Any>(javaType, qualifierAnnotations)
+    }
+    val collectionElementType = type.arguments[0].type!!
+    val elementAdapter = jsonAdapter(collectionElementType, moshi, Util.NO_ANNOTATIONS)
+    return collectionJsonAdapter(
+        elementAdapter, create, collectionElementType.isMarkedNullable) as JsonAdapter<Any?>
+  }
+
+  private inline fun <T> collectionJsonAdapter(elementAdapter: JsonAdapter<T?>,
+      crossinline create: () -> MutableCollection<T?>, nullableElements: Boolean):
+      JsonAdapter<Collection<T?>> {
+    return object : JsonAdapter<Collection<T?>>() {
```

Comment:
needs fixing: creates a new JsonAdapter for every collection property instead of using the Moshi cache.